### PR TITLE
imp (api)!: convert coins to token only once in MsgTransfer 

### PR DESCRIPTION
**Pull request by bznein**
## Description

closes: #7080

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/transfer/keeper/msg_server.go [hunk 1]
```diff
@@ -130,30 +130,22 @@ func (k Keeper) unwindHops(ctx sdk.Context, msg *types.MsgTransfer) (*types.MsgT
 // getUnwindHops returns the hops to be used during unwinding. If coins consists of more than
 // one coin, all coins must have the exact same trace, else an error is returned. getUnwindHops
 // also validates that the coins are not native to the chain.
-func (k Keeper) getUnwindHops(ctx sdk.Context, coins sdk.Coins) ([]types.Hop, error) {
+func (Keeper) getUnwindHops(tokens []types.Token) ([]types.Hop, error) {
```
[the remark this answers is not in the document]
Comment (colin-axner) [reply modules/apps/transfer/keeper/msg_server.go:133]: Makes sense to me!

**Review by colin-axner — APPROVED** [review]
Nice! Great work. The changes look good to me. It's nice to be more efficient in this approach

**Review by crodriguezvega — COMMENTED** [review]
Sorry for leaving the review after merge, but is there a reason why we couldn't merge this directly to main and then backport to v9? Then we also have these changes in the release branch. As far as I can see there is nothing in the PR that directly depends on the port router refactor, right?

File: modules/apps/transfer/keeper/export_test.go [hunk 2]
```diff
@@ -35,8 +35,8 @@ func (k Keeper) TokenFromCoin(ctx sdk.Context, coin sdk.Coin) (types.Token, erro
 }
 
 // UnwindHops is a wrapper around unwindHops for testing purposes.
-func (k Keeper) UnwindHops(ctx sdk.Context, msg *types.MsgTransfer) (*types.MsgTransfer, error) {
-	return k.unwindHops(ctx, msg)
+func (k Keeper) UnwindHops(msg *types.MsgTransfer, tokens []types.Token) (*types.MsgTransfer, error) {
```
[the remark this answers is not in the document]
Comment (bznein) [reply modules/apps/transfer/keeper/export_test.go:38]: Yeah that's a good point, I wasn't too sure about it tbh!

**Comment by bznein** [comment]
> Sorry for leaving the review after merge, but is there a reason why we couldn't merge this directly to main and then backport to v9? Then we also have these changes in the release branch. As far as I can see there is nothing in the PR that directly depends on the port router refactor, right?

You're right, I could have done it but just automatically went on "feature branch" mode :) I think it should be safe to add the same PR for main, I can do it

**Comment by bznein** [comment]
> Sorry for leaving the review after merge, but is there a reason why we couldn't merge this directly to main and then backport to v9? Then we also have these changes in the release branch. As far as I can see there is nothing in the PR that directly depends on the port router refactor, right?

@crodriguezvega actually scratch my previous comment. On main we don't have the inefficient looping (https://github.com/cosmos/ibc-go/blob/9efb7104e93ecbc1bc2e1f5e9fb770fe556633ae/modules/apps/transfer/keeper/msg_server.go#L30-L51) as we don't need to convert `coins` to `tokens` in the `Transfer` function since we use `k.sendTransfer` that takes `coins`).

(inside that function we still convert `coins` to `token` but we also use `coins` directly, so it's a bit more complex. I guess if we want this change in `main` we can make it but better to create another issue as it is definitely more involved.

**Comment by crodriguezvega** [comment]
> > Sorry for leaving the review after merge, but is there a reason why we couldn't merge this directly to main and then backport to v9? Then we also have these changes in the release branch. As far as I can see there is nothing in the PR that directly depends on the port router refactor, right?
> 
> @crodriguezvega actually scratch my previous comment. On main we don't have the inefficient looping (
> 
> https://github.com/cosmos/ibc-go/blob/9efb7104e93ecbc1bc2e1f5e9fb770fe556633ae/modules/apps/transfer/keeper/msg_server.go#L30-L51
> 
> ) as we don't need to convert `coins` to `tokens` in the `Transfer` function since we use `k.sendTransfer` that takes `coins`).
> (inside that function we still convert `coins` to `token` but we also use `coins` directly, so it's a bit more complex. I guess if we want this change in `main` we can make it but better to create another issue as it is definitely more involved.

Ah, gotcha. Thank you for explaining!

**Comment by chatton** [comment]
thanks for picking this one up, LGTM

**Pull request merged**